cpp/src/quantiles/tdigest/tdigest.cu

```diff
@@ -0,0 +1,356 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/tdigest/tdigest.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/structs/structs_column_view.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+// https://developer.nvidia.com/blog/lerp-faster-cuda/
+template <typename T>
+__device__ inline T lerp(T v0, T v1, T t)
+{
+  return fma(t, v1, fma(-t, v0, v0));
+}
+
+template <typename T>
+struct cast_output {
+  data_type t;
+
+  __device__ device_storage_type_t<T> operator()(double v)
+  {
+    if constexpr (!cudf::is_fixed_point<T>()) { return static_cast<T>(v); }
+    if constexpr (cudf::is_fixed_point<T>()) {
+      return T{v, numeric::scale_type{t.scale()}}.value();
+    }
+  }
+};
+
+// kernel for computing percentiles on input tdigest (mean, weight) centroid data.
+template <typename T>
+__global__ void compute_percentiles_kernel(offset_type const* tdigest_offsets,
+                                           size_type num_tdigests,
+                                           device_span<double const> percentiles,
+                                           double const* mean_,
+                                           double const* weight_,
+                                           double const* min_,
+                                           double const* max_,
+                                           double const* cumulative_weight_,
```

Comment:
Instead of passing in all the bits of a tdigest as independent pieces, this would be tremendously cleaner to pass in an iterator whose `value_type` is a `tdigest` object that wraps up and codifies all the pieces of a tdigest, e.g.,
```
strtuct tdigest{
   size_type size;
   double mean;
   double weight;
   double min;
   double max;
   double cumulative_weight; // is this a property of an individual tdigest?
};
```

Then you just pass in a transform iterator that constructs these objects on the fly from the relevant pointers.

Reply:
If you're worried about the cost of constructing these objects, don't be. I'd bet a shiny silver dollar that it gets optimized down into the exact same PTX.

Reply:
Based on what we talked about, I think instead of calling the struct `tdigest`, it is really called `centroid`.